src/network/protocol.zig

```diff
@@ -19,41 +25,152 @@ pub const COMMAND_SIZE: usize = 12;
 /// Magic bytes for mainnet
 pub const MAGIC_BYTES: [4]u8 = .{ 0xF9, 0xBE, 0xB4, 0xD9 };
 
+/// An IpV6 address
+pub const IpV6Address = struct {
+    ip: [8]u16, // represented in big endian
+    port: u16, // represented in system native endian
+};
+
 /// NetworkAddress represents a network address
 pub const NetworkAddress = struct {
     services: u64,
-    ip: [16]u8,
-    port: u16,
-
-    pub fn init(address: net.Address) NetworkAddress {
-        const result = NetworkAddress{
-            .services = ServiceFlags.NODE_NETWORK,
-            .ip = [_]u8{0} ** 16,
-            .port = address.getPort(),
-        };
-        // TODO: Handle untagged union properly (for IPv6)
-
-        return result;
-    }
+    address: IpV6Address,
 };
 
 /// VersionMessage represents the "version" message
+///
+/// https://developer.bitcoin.org/reference/p2p_networking.html#version
 pub const VersionMessage = struct {
     version: i32,
     services: u64,
     timestamp: i64,
     addr_recv: NetworkAddress,
-    addr_from: NetworkAddress = .{
-        .services = 0,
-        .ip = [_]u8{0} ** 16,
-        .port = 0,
-    },
-    nonce: u64 = 0,
-    user_agent: []const u8 = "",
-    start_height: i32 = 0,
-    relay: bool = false,
+    addr_trans: NetworkAddress,
+    nonce: u64,
+    user_agent: ?[]const u8,
+    start_height: i32,
+    relay: ?bool,
+
+    pub fn deinit(self: VersionMessage, allocator: std.mem.Allocator) void {
+        if (self.user_agent) |ua| {
+            allocator.free(ua);
+        }
+    }
+
+    /// Serialize a message to bytes
+    ///
+    /// The caller is responsible for freeing the returned value.
+    pub fn serialize(self: VersionMessage, allocator: std.mem.Allocator) ![]u8 {
+        // 4 + 8 + 8 + (2 * (8 + 16 + 2) + 8 + 4)
+        const fixed_length = 84;
+        var user_agent_len: usize = undefined;
+        if (self.user_agent) |ua| {
+            user_agent_len = ua.len;
+        } else {
+            user_agent_len = 0;
+        }
```

Comment:
```suggestion
        const user_agent_len = if (self.user_agent) |ua| ua.len else 0;
```

Reply:
Damn, I was looking for this exact syntax, but it is not showcased in the doc I found. Thank you

Reply:
done